File: 剑指 Offer（第 2 版）/14_2.java
/*
    和上一题的区别在于大数求余
    python 好像很适合快速幂，有a//=2 的操作
*/

class Solution {
    public int cuttingRope(int n) {
        if(n <= 3) return n - 1;
        long res = 1L;
        int p = 1000000007;
        while(n>4){
            res = (res*3)%p;
            n -= 3;
        }
        return (int)((res*n)%p);
    }
}

File: 剑指 Offer（第 2 版）/14_2.cpp
// 剑指 Offer 14- II. 剪绳子II
// https://leetcode-cn.com/problems/jian-sheng-zi-ii-lcof/


#include<vector>
#include<math.h>
using namespace std;
class Solution {
public:
    // 循环求余
    int cuttingRope(int n) {
        if(n<=3) return n-1;
        int p = 1000000007;
        long long result = 1L;
        while(n>4){
            result = (result*3)%p;
            n -= 3;
        }
        return (int)((result*n)%p);
    }
};